### PR TITLE
add SQL Schema Node Mapping + SQLTableRetrieverQueryEngine + obj index fixes

### PR DESCRIPTION
**Pull request by jerryjliu**
# Description

This PR adds a `SQLTableNodeMapping` which allows mapping a table schema to from Node, 
and a `SQLTableRetrieverQueryEngine` that can use this to dynamically retrieve table schemas.

Cleaner than previous interfaces.

Also fixed object index 

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: llama_index/objects/table_node_mapping.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            f"Schema of table {obj.table_name}:\n"
+            f"{self._sql_database.get_single_table_info(obj.table_name)}\n"
+        )
+        if obj.context_str is not None:
+            table_text += f"Context of table {obj.table_name}:\n"
+            table_text += obj.context_str
+
+        return Node(
+            text=table_text,
+            node_info={"name": obj.table_name, "context": obj.context_str},
```
Comment (logan-markewich) [inline llama_index/objects/table_node_mapping.py:54]: Any reason for using node_info here? Technically this field will be removed soon :)

File: llama_index/indices/struct_store/sql.py [hunk 2]
```diff
@@ -34,6 +34,8 @@ class SQLStructStoreIndex(BaseStructStoreIndex[SQLStructTable]):
     During query time, the user can either specify a raw SQL query
     or a natural language query to retrieve their data.
 
+    NOTE: this is deprecated.
```
Comment (logan-markewich) [inline llama_index/indices/struct_store/sql.py:37]: Future TODO: Update our structured tutorial in the docs to use the new classes

**Review by logan-markewich — APPROVED** [review]
🚢 nice, looks good! Good feature to promote for sure.

File: llama_index/objects/table_node_mapping.py (hunk 1 above)
Comment (jerryjliu) [reply llama_index/objects/table_node_mapping.py:54]: yeah...it was the main way to store information while still giving us control over what to put in node_text.

another hacky approach is just create a JSON dict and serialize to string

Comment (logan-markewich) [reply llama_index/objects/table_node_mapping.py:54]: Fair enough! Thankfully, the new schema still allows this control (just slightly more verbose I suppose)

**Pull request merged**